src/support/mtx_rw.c

```diff
@@ -245,6 +243,8 @@ stall:			__wt_cond_wait(
 	 * meantime.
 	 */
 	WT_READ_BARRIER();
+
+	WT_ASSERT(session, l->u.s.readers_active > 0);
```

Comment:
As discussed, `volatile` prevents the _compiler_ from reordering instructions, but a CPU with a weak memory model can schedule a read earlier (i.e., use a cached value).